### PR TITLE
LockedLoad

### PR DESCRIPTION
**Pull request by angelo-rendina-prima**

### CODE REVIEW

File: src/manager/locked_load.rs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+            LockedLoadInner::Some(aggregate_state) => aggregate_state,
+        }
+    }
+}
+
+/// Encapsulates the logic to avoid exposing the internal behaviour.
+/// It's essentially `Option<AggregateState<T>>`, but it also needs to
+/// retain the lock information.
+enum LockedLoadInner<T> {
+    None { id: Uuid, lock: EventStoreLockGuard },
```
Comment (cottinisimone) [inline src/manager/locked_load.rs:60]: Why do this need to have an id?

Comment (angelo-rendina-prima) [reply src/manager/locked_load.rs:60]: Because the use case is:
- you do lock_and_load(ID)
- you get Empty { ID, ... }
- you do .unwrap_or_default which gives you `AggregateState<T>::default().with_id(ID).with_lock(lock)`

TLDR: LockedLoad is the result of trying lock_and_load on a specific ID.

Comment (angelo-rendina-prima) [reply src/manager/locked_load.rs:60]: I guess there can be a scenario where you do
```
result = lock_and_load(123)
if result is None
  result = AggregateState::new() // <-- completely new ID, say 456
```

File: src/manager/locked_load.rs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+/// it contains the loaded AggregateState if found,
+/// and ensures that the lock is preserved in all cases.
+///
+/// Releases the lock on drop, unless it gets transferred through its API.
+pub struct LockedLoad<T>(LockedLoadInner<T>);
+
+impl<T> LockedLoad<T> {
+    /// Constructs a new instance from an AggregateState.
+    /// It should contain its own lock.
+    pub fn some(aggregate_state: AggregateState<T>) -> Self {
```
Comment (cottinisimone) [inline src/manager/locked_load.rs:17]:
```suggestion
    fn some(aggregate_state: AggregateState<T>) -> Self {
```

File: src/manager/locked_load.rs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        Self(LockedLoadInner::Some(aggregate_state))
+    }
+
+    /// Constructs a new instance to hold the lock for an empty state.
+    pub fn none(id: Uuid, lock: EventStoreLockGuard) -> Self {
```
Comment (cottinisimone) [inline src/manager/locked_load.rs:22]:
```suggestion
    fn none(id: Uuid, lock: EventStoreLockGuard) -> Self {
```

File: src/manager/locked_load.rs (hunk 1 above)
Comment (cottinisimone) [reply src/manager/locked_load.rs:60]: Ok, makes sense

**Review by cottinisimone — COMMENTED** [review]
Only small suggestion about constructors visibility. I don't think user should be able to create a new instance of this struct

**Review by cottinisimone — DISMISSED** [review]

File: src/manager/locked_load.rs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+        Self(LockedLoadInner::None { id, lock })
+    }
+
+    /// Checks if the AggregateState was found.
+    pub fn is_some(&self) -> bool {
+        let Self(inner) = self;
+        match inner {
+            LockedLoadInner::None { .. } => false,
+            LockedLoadInner::Some(_) => true,
+        }
```
Comment (cottinisimone) [inline src/manager/locked_load.rs:32]: What about something like?
```suggestion
        matches!(self, Self(LockedLoadInner::Some(_)))
```

**Review by cottinisimone — APPROVED** [review]
